src/libs/Performance.js

```diff
@@ -25,43 +33,157 @@ function diffObject(object, base) {
+/* eslint-enable import/no-mutable-exports */
 
-    const performance = require('react-native-performance').default;
-    const PerformanceObserver = require('react-native-performance').PerformanceObserver;
-    new PerformanceObserver((list) => {
-        if (list.getEntries().find(entry => entry.name === 'nativeLaunchEnd')) {
-            performance.measure('nativeLaunch', 'nativeLaunchStart', 'nativeLaunchEnd');
-
-            // eslint-disable-next-line no-undef
-            if (__DEV__) {
-                performance.measure('jsBundleDownload', 'downloadStart', 'downloadEnd');
-            } else {
-                performance.measure('runJsBundle', 'runJsBundleStart', 'runJsBundleEnd');
-            }
-        }
-    }).observe({type: 'react-native-mark', buffered: true});
-}
+if (canCapturePerformanceMetrics()) {
+    /**
+     * Sets up an observer to capture events recorded in the native layer before the app fully initializes.
+     */
+    setupPerformanceObserver = () => {
+        const performanceReported = require('react-native-performance-flipper-reporter');
+        performanceReported.setupDefaultFlipperReporter();
 
-/**
- * Outputs performance stats. We alert these so that they are easy to access in release builds.
- */
-function printPerformanceMetrics() {
-    const performance = require('react-native-performance').default;
-    const entries = _.map(performance.getEntriesByType('measure'), entry => ({
-        name: entry.name, duration: Math.floor(entry.duration),
-    }));
-    alert(JSON.stringify(entries, null, 4));
+        const perfModule = require('react-native-performance');
+        perfModule.setResourceLoggingEnabled(true);
+        performance = perfModule.default;
+
+        // Monitor some native marks that we want to put on the timeline
+        new perfModule.PerformanceObserver((list, observer) => {
+            list.getEntries()
+                .forEach((entry) => {
+                    if (entry.name === 'nativeLaunchEnd') {
+                        performance.measure('nativeLaunch', 'nativeLaunchStart', 'nativeLaunchEnd');
+                    }
+                    if (entry.name === 'downloadEnd') {
+                        performance.measure('jsBundleDownload', 'downloadStart', 'downloadEnd');
+                    }
+                    if (entry.name === 'runJsBundleEnd') {
+                        performance.measure('runJsBundle', 'runJsBundleStart', 'runJsBundleEnd');
+                    }
+
+                    // We don't need to keep the observer past this point
+                    if (entry.name === 'runJsBundleEnd' || entry.name === 'downloadEnd') {
+                        observer.disconnect();
+                    }
+                });
+        }).observe({type: 'react-native-mark', buffered: true});
+
+        // Monitor for "_end" marks and capture "_start" to "_end" measures
+        new perfModule.PerformanceObserver((list) => {
```

Comment:
I haven't used observers too much, but is there any reason to not use the `PerformanceObserver` we already created above to do this?

Reply:
The observer above receives only `native` (non js marks) -  `{type: 'react-native-mark', buffered: true}`

This observer receives only "App" (js) marks

We could use a single observer that receives events from all types, but such observers cannot be configured as `buffered: true` and we want that because
> Passing buffered: true would include entries produced before the observe() call which is useful to delay handing of measurements until after performance critical startup processing.

Reply:
Got it. I missed that part. Makes sense to me.